src/context/values.js

```diff
@@ -0,0 +1,766 @@
+        if (inputValues) {
+          // The node specifies its own input values and they need to be
+          // recomputed with parent and dep values.
+          newValue = callRecursiveCompute(
+            compute,
+            node,
+            inputValues,
+            parentValue,
+            depValues
+          );
+        } else if (isDefined(parentValue)) {
+          // The node doesn't specify its own value, but parent is available.
+          // Since parent is available, it means that the node is recursive.
+          newValue = parentValue;
+        }
+      } else if (compute) {
+        newValue = callCompute(
+          compute,
+          node,
+          inputValues || EMPTY_ARRAY,
+          depValues
+        );
+      }
+    }
+
+    return newValue;
+  }
+
+  /**
+   * Update the node from which the parent value is used.
+   *
+   * @param {!UsedDef} used
+   * @param {?./node.ContextNode} newParentValueNode
+   * @private
+   */
+  updateParentValueNode_(used, newParentValueNode) {
+    const {prop, parentValueNode: oldParentValueNode, pingParent} = used;
+    if (newParentValueNode != oldParentValueNode) {
+      used.parentValueNode = newParentValueNode;
+      used.parentValue = undefined;
+
+      if (oldParentValueNode) {
+        oldParentValueNode.values.unsubscribe(prop, devAssert(pingParent));
+      }
+
+      if (newParentValueNode) {
+        newParentValueNode.values.subscribe(prop, devAssert(pingParent));
+      }
+    }
+  }
+}
+
+/**
+ * See `Values.scan()` method.
+ *
+ * @param {!./node.ContextNode} contextNode
+ * @param {!ContextProp} prop
+ * @return {boolean}
+ */
+function scan(contextNode, prop) {
+  return contextNode.values.scan(prop);
+}
+
+/**
+ * See `Values.scanAll()` method.
+ *
+ * @param {!./node.ContextNode} contextNode
+ * @param {*} unusedArg
+ * @param {!Array<string>} state
+ * @return {!Array<string>}
+ */
+function scanAll(contextNode, unusedArg, state) {
+  return contextNode.values.scanAll(state);
+}
+
+/**
+ * See `Values.has()` method.
+ *
+ * @param {!./node.ContextNode} contextNode
+ * @param {!ContextProp} prop
+ * @return {boolean}
+ */
+function hasInput(contextNode, prop) {
+  return contextNode.values.has(prop);
+}
+
+/**
+ * Whether the property is recursive.
+ *
+ * @param {!ContextProp} prop
+ * @return {boolean}
+ */
+function isRecursive(prop) {
+  // Only `false` values make a value non-recursive. `true` and
+  // `function` values are considered recursive.
+  return !!prop.needsParent;
+}
+
+/**
+ * Whether the parent value is required to calculate the used value.
+ *
+ * @param {!ContextProp} prop
+ * @param {?Array} inputs
+ * @return {boolean}
+ */
+function calcNeedsParent(prop, inputs) {
+  const {needsParent, compute} = prop;
+  if (typeof needsParent == 'function') {
+    return inputs ? needsParent(inputs) : true;
+  }
+  if (needsParent && inputs && !compute) {
+    // The default `compute` function is OR. Thus, when inputs are specified,
```

Comment:
nit: is "OR" the right word here? The conclusion makes sense to me though. default compute fn when inputs are provided is to take `inputs[0]`, which is independent of the parent.

Reply:
changed.